client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java

```diff
@@ -522,11 +530,212 @@ protected void set(K key, V value) {
+               // Convert to <K, V> for user
+               e -> new AbstractMap.SimpleImmutableEntry<>(e.getKey(), e.getValue().getValue()));
+      }
+
+      @Override
+      public CloseableSpliterator<Entry<K, V>> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT);
+      }
+
+      @Override
+      public Stream<Entry<K, V>> stream() {
+         return Closeables.stream(spliterator(), false);
+      }
+
+      @Override
+      public Stream<Entry<K, V>> parallelStream() {
+         return Closeables.stream(spliterator(), true);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public void clear() {
+         RemoteCacheImpl.this.clear();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         Map.Entry entry = toEntry(o);
+         if (entry != null) {
+            V value = RemoteCacheImpl.this.get(entry.getKey());
+            return value != null && value.equals(entry.getValue());
+         }
+         return false;
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         Map.Entry<K, V> entry = toEntry(o);
+         return entry != null && RemoteCacheImpl.this.removeEntry(entry);
+      }
+
+      private Map.Entry<K, V> toEntry(Object obj) {
+         if (obj instanceof Map.Entry) {
+            return (Map.Entry) obj;
+         } else {
+            return null;
+         }
+      }
+   }
+
+   private class ValuesCollection extends AbstractCollection<V> implements CloseableIteratorCollection<V> {
+
+      @Override
+      public CloseableIterator<V> iterator() {
+         return new CloseableIteratorMapper<>(new RemovableCloseableIterator<>(
+               // First make <K, VersionedValue<V>>
+               castIterator(retrieveEntriesWithMetadata(null, batchSize)),
+               // Apply remove using version
+               e -> removeWithVersion(e.getKey(), e.getValue().getVersion())),
+               // Convert to V for user
+               e -> e.getValue().getValue());
+      }
+
+      @Override
+      public CloseableSpliterator<V> spliterator() {
+         return Closeables.spliterator(iterator(), Long.MAX_VALUE, Spliterator.NONNULL | Spliterator.CONCURRENT);
+      }
+
+      @Override
+      public Stream<V> stream() {
+         return Closeables.stream(spliterator(), false);
+      }
+
+      @Override
+      public Stream<V> parallelStream() {
+         return Closeables.stream(spliterator(), true);
+      }
+
+      @Override
+      public int size() {
+         return RemoteCacheImpl.this.size();
+      }
+
+      @Override
+      public void clear() {
+         RemoteCacheImpl.this.clear();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         // Have to close the stream, just in case stream terminates early
```

Comment:
This is another one that could be more efficient if implemented as a remote server task?